src/main/java/org/broadinstitute/hellbender/utils/hmm/ForwardBackwardAlgorithm.java

```diff
@@ -393,33 +442,107 @@ public double logProbability(final S state, final T position) {
+                    throw new IllegalArgumentException("the input position " + position + " is nowhere to be found in the original position sequence");
+                } else if (lastIndex >= data.size()) {
+                    throw new IllegalArgumentException("the input state sequence is too long");
+                }
+                double result = logForwardProbability(positions.get(startIndex), states.get(0));
+                for (int statesOffset = 1, dataOffset = startIndex + 1; dataOffset <= lastIndex; statesOffset++, dataOffset++) {
+                    result += model.logTransitionProbability(states.get(statesOffset - 1), positions.get(dataOffset - 1), states.get(statesOffset), positions.get(dataOffset));
+                    result += model.logEmissionProbability(data.get(dataOffset), states.get(statesOffset), positions.get(dataOffset));
                 }
-                result -= logDataLikelihood;
+                result += logBackwardProbability(positions.get(lastIndex), states.get(statesLength - 1));
+                result -= logDataLikelihood[lastIndex];
                 return result;
             }
         }
 
         @Override
+        public double logConstrainedProbability(final T position, final List<Set<S>> stateConstraints) {
+            Utils.nonNull(stateConstraints, "the input state constraints sequence cannot be null");
+            if (stateConstraints.size() == 0) {
+                return 0;
+            } else {
+                final int length = stateConstraints.size();
+                final int startIndex = positionIndex.get(position);
+                final int lastIndex = length + startIndex - 1;
+                if (startIndex == -1) {
+                    throw new IllegalArgumentException("the input position " + position + " is nowhere to be found in the original position sequence");
+                } else if (lastIndex >= data.size()) {
+                    throw new IllegalArgumentException("the input state sequence is too long");
+                }
+                List<S>  currentStates = new ArrayList<>(Utils.nonNull(stateConstraints.get(0)));
+                double[] currentLikelihoods = currentStates.stream()
+                        .mapToInt(stateIndex::getInt)
+                        .mapToDouble(i -> logForwardProbabilities[startIndex][i])
+                        .toArray();
+                for (int statesOffset = 1, dataOffset = startIndex + 1; dataOffset <= lastIndex; statesOffset++, dataOffset++) {
+                    final double[] previousLikelihoods = currentLikelihoods;
+                    final List<S> previousStates = currentStates;
+                    final T previousPosition = positions.get(dataOffset - 1);
+                    final T thisPosition = positions.get(dataOffset);
+                    final D thisData = data.get(dataOffset);
+
+                    currentStates = new ArrayList<>(Utils.nonNull(stateConstraints.get(statesOffset)));
```

Comment:
Put in a comment about how you're considering all hidden state transitions from the previous position's constrained set to the current position's.